bugtool/cmd/root.go

```diff
@@ -0,0 +1,237 @@
+		{"/proc/config.gz", fmt.Sprintf("%s/kernel-config.gz", confDir)},
+		{fmt.Sprintf("/boot/config-%s", kernel), fmt.Sprintf("%s/kernel-config-%s", confDir, kernel)},
+	}
+
+	for _, location := range locations {
+		if _, err := os.Stat(location.Src); os.IsNotExist(err) {
+			continue
+		}
+		err := copyFile(location.Src, location.Dst)
+		if err != nil {
+			log.WithError(err).Warn("could not copy kernel config")
+		}
+	}
+}
+
+func copySystemCmds(cmdDir string) {
+	// Not expecting all of the commands to be available.
+	sources := []string{
+		// Host and misc
+		"hostname", "ip a", "ip r", "ip link", "iptables-save", "uname -a",
+		"dig", "netstat", "pidstat", "arp", "top -b -n 1",
+		"uptime", "dmesg", "bpftool map show", "bpftool prog show",
+		"iptables -S", "ip6tables -S", "iptables -L -v", "ps",
+		"cat /proc/sys/net/core/bpf_jit_enable", "cat /proc/kallsyms",
+		// Versions
+		"docker version",
+		"kubectl version",
+		// Try to find some useful logs
+		"cat /var/log/upstart/docker.log",
+		"cat /var/log/docker.log", "cat /var/log/daemon.log",
+		"journalctl -u cilium*", "cat /var/log/messages", "journalctl -u kubelet",
+		"kubectl -n kube-system get pods",
+		"kubectl get pods,svc --all-namespaces",
+		"kubectl get version",
+	}
+
+	// Append ethtool links
+	if links, err := netlink.LinkList(); err == nil {
+		for _, link := range links {
+			// query current settings
+			sources = append(sources, fmt.Sprintf("ethtool %s", link.Attrs().Name))
+			// query for driver information
+			sources = append(sources, fmt.Sprintf("ethtool -i %s", link.Attrs().Name))
+		}
+	}
+
+	execs := make(chan bool)
+	runInBackground := func(dir, cmd string) {
+		writeCmdToFile(dir, cmd)
+		execs <- true
+	}
+	for _, cmd := range sources {
+		go runInBackground(cmdDir, cmd)
+	}
+	// Wait for the commands to complete
+	for i := 0; i < len(sources); i++ {
+		<-execs
+	}
+}
+
+func copyCiliumCmds(ciliumDir string) {
+	sources := []string{
+		// Most of the output should come via debuginfo but also adding the other ones for skimming purposes
+		"cilium debuginfo", "cilium config", "cilium bpf tunnel list", "cilium bpf lb list",
+		"cilium bpf endpoint list", "cilium bpf ct list global",
+	}
+
+	for _, cmd := range sources {
+		// Add the host flag if set
+		if len(host) > 0 {
+			cmd = fmt.Sprintf("%s -H %s", cmd, host)
+		}
+		writeCmdToFile(ciliumDir, cmd)
+	}
+}
+
+func execCommand(cmd string, args ...string) (string, error) {
+	fmt.Printf("exec: %s %s\n", cmd, args)
+	output, err := exec.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		return string(output), nil
+	}
+	return string(output), nil
+}
+
+// writeCmdToFile will execute command and write markdown output to a file
+func writeCmdToFile(cmdDir, prompt string) {
+	// Clean up the filename
+	name := strings.Replace(prompt, "/", " ", -1)
+	name = strings.Replace(name, " ", "-", -1)
+	f, _ := os.Create(filepath.Join(cmdDir, name+".md"))
```

Comment:
It might be better to exit on error, since f is undefined in that case (or nil) and that might cause weird panics later when we try to use f.